### ISSUE TITLE
idea: Precision scaling research

### ISSUE THREAD

**Issue by hahuyhoang411** [janhq/ichigo#127]
### Problem Statement

Hypothesis: Increasing numerical precision during training can improve the performance of small language models (≈1B parameters), potentially enabling them to achieve capabilities comparable to larger models (3B-7B parameters).

### Idea

Reference: https://arxiv.org/pdf/2411.04330

**Comment by hahuyhoang411** [comment]
@bachvudinh as co-author

We have ran some testing to train [Llama 3.2 1B Instruct](https://huggingface.co/meta-llama/Llama-3.2-1B-Instruct) to check if `fp32` can perform better `bf16`


Precision | Learning Rate | Weight Decay | Global batchsize | Trained Samples | Final Loss | Loss curve | MMLU
-- | -- | -- | -- | -- | -- | -- | --
fp32 | 3e-4 | 0.01 | 96 | 0.2M | 1.24 |   |  
fp32 | 2.5e-4 | 0.01 | 96 | 0.2M | 1.22 |   |  
bf16 | 3e-4 | 0.01 | 96 | 0.2M | exploded |   |  
bf16 | 2e-4 | 0.01 | 96 | 0.2M | exploded |   |  
b16 | 2.5e-4 | 0.01 | 96 | 0.2M | 1.26 |   |  
fp32 | 3e-4 | 0.2 | ? | 0.5M | 0.67 |   |  
fp32 | 1e-4 | 0.05 | ? | 1.7M | 1.32 |   |

**Comment by hahuyhoang411** [comment]
fp32 3e-4
![Screenshot 2024-11-21 at 01.09.02.png](https://prod-files-secure.s3.us-west-2.amazonaws.com/e8320b38-242d-4a8e-a3a8-6fd4aad5d470/652675f0-da40-4871-9369-706c80446bce/Screenshot_2024-11-21_at_01.09.02.png)

**Comment by hahuyhoang411** [comment]
fp32 2.5e-4
<img width="616" alt="Screenshot 2024-11-21 at 01 09 49" src="https://github.com/user-attachments/assets/d98540fa-1c4a-4415-a6aa-db36ff954757">

**Comment by hahuyhoang411** [comment]
fp32 3e-4
<img width="1037" alt="Screenshot 2024-11-21 at 01 19 44" src="https://github.com/user-attachments/assets/cc0f65c1-ada8-4ae4-90ea-633271dd84a1">

**Comment by hahuyhoang411** [comment]
fp32 1.7M
<img width="1038" alt="Screenshot 2024-11-21 at 01 20 14" src="https://github.com/user-attachments/assets/1cf4bbcc-8526-4ddb-8b76-f832aedb2eaa">

**Comment by tikikun** [comment]
A few pending issues:
- It's very obvious that even tho more stabilized the training is not converging and still have hiccup, which indicating issue inside the optimizer itself not being able to give direction good enough for the optimizing process
- Since we are currently hitting a wall on the optimizer itself, we will not continue scaling up the precision

Next steps:
- @tikikun  to do his own research on optimizer
- We make use of the cluster for training qwen 32b instruct 

cc @0xSage for interested
